apps/pipelines/nodes/nodes.py

```diff
@@ -148,40 +213,54 @@ def _create_model_from_data(value_data, model_name: str):
         return schema
 
 
-class UpdateParticipantMemory(PipelineNode):
-    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
-    the input data will be merged with `key_name` as the key name.
-    """
+class ExtractStructuredData(ExtractStructuredDataNodeMixin, LLMResponse):
+    __human_name__ = "Extract Structured Data"
+    data_schema: str
+
 
-    __human_name__ = "Update participant memory"
+class ExtractParticipantData(ExtractStructuredDataNodeMixin, LLMResponse):
+    __human_name__ = "Extract Participant Data"
+    data_schema: str
     key_name: str | None = None
 
-    def _process(self, state: PipelineState) -> PipelineState:
-        """Strategy:
-        - If the input is a string or list, a `key_name` must be present to be valid JSON.
-        - If `key_name` is present, we create a new dictionary with `input` as the value to `key_name`.
-        - Merge this new dictionary with the current participant data, essentially overriding existing data
-        """
-        new_data = state["messages"][-1]
+    def get_json_schema(self):
+        return ExtractParticipantData.to_json_schema(json.loads(self.data_schema))
+
+    def get_reference_data(self, state) -> dict:
         session = state["experiment_session"]
+        data = session.get_participant_data()
+        if self.key_name:
+            return data.get(self.key_name, {})
+        return data
 
+    def update_reference_data(self, new_data: dict, reference_data: dict) -> dict:
         if isinstance(new_data, str) or isinstance(new_data, list):
             if not self.key_name:
                 raise KeyError("A key is expected for a string or list value.")
 
+        if isinstance(new_data, str):
+            return new_data
+
+        if isinstance(new_data, list):
+            return new_data
+
+        return reference_data | new_data
+
+    def post_extraction_hook(self, output, state):
+        session = state["experiment_session"]
         if self.key_name:
-            new_data = {self.key_name: new_data}
+            output = {self.key_name: output}
         try:
             participant_data = ParticipantData.objects.for_experiment(session.experiment).get(
                 participant=session.participant
             )
 
-            participant_data.data = participant_data.data | new_data
+            participant_data.data = participant_data.data | output
             participant_data.save()
         except ParticipantData.DoesNotExist:
             ParticipantData.objects.create(
                 participant=session.participant,
                 content_object=session.experiment,
                 team=session.team,
-                data=new_data,
+                data=output,
             )
```

Comment:
this try except could be replaced with by using `ParticipantData.objects.update_or_create` e.g. https://github.com/dimagi/open-chat-studio/blob/2ed3328579dff3f9685183e5e89b1a5ef2ef97d2/apps/api/views.py#L131-L137

Reply:
I don't think this will work here, because when the `key_name` is specified, the `output` will be a subset of the participant data (more specificially, the subset corresponding to `key_name`), in which case we need to fetch -> merge -> save.

Reply:
Wait, I think I'm wrong